src/qibolab/instruments/zhinst/sweep.py

```
@@ -0,0 +1,139 @@
+        pulse_sweeps = []
+        channel_sweeps = []
+        parallel_sweeps = []
+        for sweeper in sweepers:
+            for pulse in sweeper.pulses or []:
+                if sweeper.parameter in (Parameter.duration, Parameter.start):
+                    sweep_param = lo.SweepParameter(
+                        values=sweeper.values * NANO_TO_SECONDS
+                    )
+                    pulse_sweeps.append((pulse, sweeper.parameter, sweep_param))
+                elif sweeper.parameter is Parameter.frequency:
+                    ptype, qubit = pulse.type, qubits[pulse.qubit]
+                    if ptype is PulseType.READOUT:
+                        ch = measure_channel_name(qubit)
+                        intermediate_frequency = (
+                            qubit.readout_frequency
+                            - qubit.readout.local_oscillator.frequency
+                        )
+                    elif ptype is PulseType.DRIVE:
+                        ch = qubit.drive.name
+                        intermediate_frequency = (
+                            qubit.drive_frequency
+                            - qubit.drive.local_oscillator.frequency
+                        )
+                    else:
+                        raise ValueError(
+                            f"Cannot sweep frequency of pulse of type {ptype}, because it does not have associated frequency"
+                        )
+                    sweep_param = lo.SweepParameter(
+                        values=sweeper.values + intermediate_frequency
+                    )
+                    channel_sweeps.append((ch, sweeper.parameter, sweep_param))
+                elif (
+                    pulse.type is PulseType.READOUT
+                    and sweeper.parameter is Parameter.amplitude
+                ):
+                    sweep_param = lo.SweepParameter(
+                        values=sweeper.values / max(sweeper.values)
```

Do you understand why a normalization is needed? (and only in this specific case)

---

For the readout amplitude I have not figured out it yet. I just kept it here to comply with the previous behavior. For other pulses, like drive, some normalization is done and the reason for that is the laboneq amplitude sweeper does not accept values larger than 1

---

But then, shouldn't we keep the error transparent to the user?

Here, it seems to me that the information about `max(sweeper.values)` is silently lost.

---

I was checking, it does take values greater than 1, but it just trims them back to 1 and logs this. I think for the readout I followed the same logic as with the drive pulses. This was due to the `FACTOR` sweeper for the amplitudes.